File: 프로그래머스/1/42840. 모의고사/모의고사.py
def solution(answers):
    pattern = [[1, 2, 3, 4, 5], [2, 1, 2, 3, 2, 4, 2, 5], [3, 3, 1, 1, 2, 2, 4, 4, 5, 5]]
    
    n = len(answers)
    result = {}
    for i, p in enumerate(pattern):
        cnt = 0
        for j in range(n):
            if answers[j] == p[j % len(p)]:
                cnt += 1
        result[i + 1] = cnt
    
    max_value = max(result.values())
    return [i + 1 for i in range(3) if result[i + 1] == max_value]

File: 프로그래머스/1/42840. 모의고사/모의고사.java
import java.util.*;

class Solution {
    public int[] solution(int[] answers) {
        int[][] pattern = {{1, 2, 3, 4, 5}, {2, 1, 2, 3, 2, 4, 2, 5}, {3, 3, 1, 1, 2, 2, 4, 4, 5, 5}};
        
        int n = answers.length;
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < 3; i++) {
            int cnt = 0;
            for (int j = 0; j < n; j++) {
                if (answers[j] == pattern[i][j % pattern[i].length]) {
                    cnt += 1;
                }    
            }
            map.put(i + 1, cnt);
        }
        
        int max = Collections.max(map.values());
        
        List<Integer> list = new ArrayList<>();
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if (max == entry.getValue()) {
                list.add(entry.getKey());
            }
        }
        
        int[] answer = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i);
        }
        return answer;
    }
}